```
==============================
Failed 1 tests - output below:
==============================

nova.tests.unit.privsep.test_libvirt.LibvirtTestCase.test_readpty
-----------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File "/home/i500815/src/nova/nova/tests/unit/privsep/test_libvirt.py", line 145, in test_readpty
    mock_fcntl.fcntl.assert_has_calls(

      File "/usr/lib/python3.10/unittest/mock.py", line 966, in assert_has_calls
    raise AssertionError(

    AssertionError: Calls not found.
Expected: [call(<_AutospecMagicMock name='open().fileno()' id='125384671543600'>, <_AutospecMock name='mock.F_GETFL' id='125385496855744'>),
 call(<_AutospecMagicMock name='open().fileno()' id='125384671543600'>, <_AutospecMock name='mock.F_SETFL' id='125385496861168'>, 34817)]


Captured pythonlogging:
~~~~~~~~~~~~~~~~~~~~~~~
    2026-03-12 12:04:19,135 WARNING [oslo_policy.policy] JSON formatted policy_file support is deprecated since Victoria release. You need to use YAML format which will be default in future. You can use ``oslopolicy-convert-json-to-yaml`` tool to convert existing JSON-formatted policy file to YAML-formatted in backward compatible way: https://docs.openstack.org/oslo.policy/latest/cli/oslopolicy-convert-json-to-yaml.html.
2026-03-12 12:04:19,136 WARNING [oslo_policy.policy] JSON formatted policy_file support is deprecated since Victoria release. You need to use YAML format which will be default in future. You can use ``oslopolicy-convert-json-to-yaml`` tool to convert existing JSON-formatted policy file to YAML-formatted in backward compatible way: https://docs.openstack.org/oslo.policy/latest/cli/oslopolicy-convert-json-to-yaml.html.
2026-03-12 12:04:19,139 WARNING [oslo_policy.policy] Policy Rules ['os_compute_api:extensions', 'os_compute_api:os-floating-ip-pools', 'os_compute_api:os-quota-sets:defaults', 'os_compute_api:os-availability-zone:list', 'os_compute_api:limits', 'project_member_api', 'project_reader_api', 'project_member_or_admin', 'project_reader_or_admin', 'os_compute_api:limits:other_project', 'os_compute_api:os-lock-server:unlock:unlock_override', 'os_compute_api:os-quota-class-sets:create', 'os_compute_api:sap:endpoints:list', 'os_compute_api:sap:clear-quota-resources-cache', 'os_compute_api:sap:in-cluster-vmotion', 'os_compute_api:sap:all-instance-uuids', 'os_compute_api:sap:usage-by-az', 'os_compute_api:sap:get-scheduler-settings', 'os_compute_api:os-server-groups:update', 'os_compute_api:servers:create:zero_disk_flavor', 'compute:servers:resize:cross_cell', 'os_compute_api:os-shelve:unshelve_to_host'] specified in policy files are the same as the defaults provided by the service. You can remove these rules from policy files which will make maintenance easier. You can detect these redundant rules by ``oslopolicy-list-redundant`` tool also.
2026-03-12 12:04:19,159 INFO [nova.privsep.libvirt] Ignored error while reading from instance console pty: module 'select' has no attribute 'epoll'
```

unit tests fail for me locally with `tox -e unit`
the "select has no attribute epoll" is weird. Python 3.10 on Ubuntu 22.04
but the other one seems valid, since we removed all the old code.

--- a/nova/conf/libvirt.py
+++ b/nova/conf/libvirt.py
@@ -104,7 +104,7 @@ Related options:
 """),
     cfg.StrOpt('virt_type',
                default='kvm',
-               choices=('kvm', 'lxc', 'qemu', 'parallels'),
+               choices=('kvm', 'lxc', 'qemu', 'parallels', 'ch'),
                help="""
 Describes the virtualization type (or so called domain type) libvirt should
 use.

--- a/nova/conf/libvirt.py
+++ b/nova/conf/libvirt.py
@@ -274,6 +274,7 @@ in following list:
 * 'kvm': 'qemu+tcp://%s/system'
 * 'qemu': 'qemu+tcp://%s/system'
 * 'parallels': 'parallels+tcp://%s/system'
+* 'ch': 'ch+tcp://%s/system'
 
 Related options:
 

--- a/nova/privsep/libvirt.py
+++ b/nova/privsep/libvirt.py
@@ -156,15 +156,20 @@ def readpty(path):
     # exception here... Some platforms (I'm looking at you Windows)
     # don't have a fcntl and we may as well let them know that
     # with an ImportError, not that they should be calling this at all.
-    import fcntl
+    import select
 
     try:
+        epoll = select.epoll()
         with open(path, 'r') as f:
-            current_flags = fcntl.fcntl(f.fileno(), fcntl.F_GETFL)
-            fcntl.fcntl(f.fileno(), fcntl.F_SETFL,
-                        current_flags | os.O_NONBLOCK)
-
-            return f.read()
+            os.set_blocking(f.fileno(), False)
+            epoll.register(f.fileno(), select.EPOLLIN)
+            poll_list = epoll.poll(1)
+            data = ''
+            for _ in poll_list:
+                data += f.read()
+            epoll.unregister(f.fileno())
+            epoll.close()
+            return data
 
     except Exception as exc:
         # NOTE(mikal): dear internet, I see you looking at me with your

--- a/nova/virt/libvirt/blockinfo.py
+++ b/nova/virt/libvirt/blockinfo.py
@@ -95,6 +95,7 @@ SUPPORTED_DEVICE_BUSES = {
     'qemu': ['virtio', 'scsi', 'ide', 'usb', 'fdc', 'sata'],
     'kvm': ['virtio', 'scsi', 'ide', 'usb', 'fdc', 'sata'],
     'lxc': ['lxc'],
+    'ch': ['virtio'],
     'parallels': ['ide', 'scsi'],
     # we no longer support UML or Xen, but we keep track of their bus types so
     # we can reject them for other virt types
@@ -276,6 +277,8 @@ def get_disk_bus_for_device_type(instance,
             return "ide"
         elif device_type == "disk":
             return "scsi"
+    elif virt_type == "ch":
+        return "virtio"
     else:
         # If virt-type not in list then it is unsupported
         raise exception.UnsupportedVirtType(virt=virt_type)

--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -6807,9 +6807,7 @@ class LibvirtDriver(driver.ComputeDriver):
                 guest_cfg, vconfig.LibvirtConfigGuestConsole,
                 log_path=log_path)
         elif CONF.libvirt.virt_type == "ch":
-            consolepty = vconfig.LibvirtConfigGuestSerial()
-            consolepty.type = "pty"
-            guest_cfg.add_device(consolepty)
+            self._create_consoles_ch(guest_cfg, instance, flavor, image_meta)
 
         else:  # qemu, kvm
             if self._is_s390x_guest(image_meta):
@@ -6838,6 +6836,29 @@ class LibvirtDriver(driver.ComputeDriver):
     def _is_x86_guest(self, image_meta: 'objects.ImageMeta') -> bool:
         archs = (fields.Architecture.I686, fields.Architecture.X86_64)
         return self._check_emulation_arch(image_meta) in archs
+
+    def _create_consoles_ch(self, guest_cfg, instance, flavor,
+                            image_meta):
+        """Create a serial device definition for a CH guest.
+
+        If `CONF.serial_console` is enabled, the web serial console feature is
+        enabled. In this case, we configure a serial device in TCP mode, in
+        order to allow the nova-serialproxy to forward the serial connection to
+        the web serial console.
+        Per default, a serial device in PTY mode is used.
+        """
+        char_dev_cls = vconfig.LibvirtConfigGuestSerial
+        log_path = self._get_console_log_path(instance)
+        if CONF.serial_console.enabled:
+            if not self._serial_ports_already_defined(instance):
+                num_ports = hardware.get_number_of_serial_ports(flavor,
+                                                                image_meta)
+                self._check_number_of_serial_console(num_ports)
+                self._create_serial_consoles(guest_cfg, num_ports,
+                                             char_dev_cls, log_path)
+        else:
+            self._create_pty_device(guest_cfg, char_dev_cls,
+                                    log_path=log_path)
 
     def _create_consoles_qemu_kvm(self, guest_cfg, instance, flavor,
                                   image_meta):

--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -1379,6 +1379,8 @@ class LibvirtDriver(driver.ComputeDriver):
             uri = CONF.libvirt.connection_uri or 'lxc:///'
         elif CONF.libvirt.virt_type == 'parallels':
             uri = CONF.libvirt.connection_uri or 'parallels:///system'
+        elif CONF.libvirt.virt_type == 'ch':
+            uri = CONF.libvirt.connection_uri or 'ch:///session'
         else:
             uri = CONF.libvirt.connection_uri or 'qemu:///system'
         return uri
@@ -4494,7 +4496,8 @@ class LibvirtDriver(driver.ComputeDriver):
         path_sources = [
             ('file', "./devices/console[@type='file']/source[@path]", 'path'),
             ('tcp', "./devices/console[@type='tcp']/log[@file]", 'file'),
-            ('pty', "./devices/console[@type='pty']/source[@path]", 'path')]
+            ('pty', "./devices/console[@type='pty']/source[@path]", 'path'),
+            ('pty', "./devices/serial[@type='pty']/source[@path]", 'path'), ]
         console_type = ""
         console_path = ""
         for c_type, epath, attrib in path_sources:
@@ -6749,6 +6752,9 @@ class LibvirtDriver(driver.ComputeDriver):
             guest.os_init_path = "/sbin/init"
             guest.os_cmdline = CONSOLE
             guest.os_init_env["product_name"] = "OpenStack Nova"
+        elif CONF.libvirt.virt_type == "ch":
+            guest.virt_type = 'kvm'
+            guest.os_kernel = "/var/lib/nova/hypervisor-fw"
         elif CONF.libvirt.virt_type == "parallels":
             if guest.os_type == fields.VMMode.EXE:
                 guest.os_init_path = "/sbin/init"
@@ -6794,6 +6800,11 @@ class LibvirtDriver(driver.ComputeDriver):
             self._create_pty_device(
                 guest_cfg, vconfig.LibvirtConfigGuestConsole,
                 log_path=log_path)
+        elif CONF.libvirt.virt_type == "ch":
+            consolepty = vconfig.LibvirtConfigGuestSerial()
+            consolepty.type = "pty"
+            guest_cfg.add_device(consolepty)
+
         else:  # qemu, kvm
             if self._is_s390x_guest(image_meta):
                 self._create_consoles_s390x(
@@ -6983,6 +6994,10 @@ class LibvirtDriver(driver.ComputeDriver):
         here explicitly so that we can _disable_ it (by setting the model to
         'none') if it's not necessary.
         """
+        # CH does not support emulation of USB controllers currently
+        if CONF.libvirt.virt_type == "ch":
+            return
+
         usbhost = vconfig.LibvirtConfigGuestUSBHostController()
         usbhost.index = 0
         # an unset model means autodetect, while 'none' means don't add a
@@ -7391,6 +7406,8 @@ class LibvirtDriver(driver.ComputeDriver):
     @staticmethod
     def _guest_add_video_device(guest):
         if CONF.libvirt.virt_type == 'lxc':
+            return False
+        elif CONF.libvirt.virt_type == "ch":
             return False
 
         # NB some versions of libvirt support both SPICE and VNC

--- a/nova/virt/libvirt/driver.py
+++ b/nova/virt/libvirt/driver.py
@@ -1380,7 +1380,7 @@ class LibvirtDriver(driver.ComputeDriver):
         elif CONF.libvirt.virt_type == 'parallels':
             uri = CONF.libvirt.connection_uri or 'parallels:///system'
         elif CONF.libvirt.virt_type == 'ch':
-            uri = CONF.libvirt.connection_uri or 'ch:///session'
+            uri = CONF.libvirt.connection_uri or 'ch:///system'
         else:
             uri = CONF.libvirt.connection_uri or 'qemu:///system'
         return uri
@@ -1391,6 +1391,7 @@ class LibvirtDriver(driver.ComputeDriver):
             'kvm': 'qemu+%(scheme)s://%(dest)s/system',
             'qemu': 'qemu+%(scheme)s://%(dest)s/system',
             'parallels': 'parallels+tcp://%(dest)s/system',
+            'ch': 'ch+%(scheme)s://%(dest)s/system',
         }
         dest = oslo_netutils.escape_ipv6(dest)
 

--- a/nova/virt/libvirt/vif.py
+++ b/nova/virt/libvirt/vif.py
@@ -78,6 +78,9 @@ SUPPORTED_VIF_MODELS = {
         network_model.VIF_MODEL_RTL8139,
         network_model.VIF_MODEL_E1000,
     ],
+    'ch': [
+        network_model.VIF_MODEL_VIRTIO,
+    ],
 }
 
 
@@ -172,7 +175,7 @@ class LibvirtGenericVIFDriver(object):
         # If the virt type is KVM/QEMU/VZ(Parallels), then use virtio according
         # to the global config parameter
         if (model is None and CONF.libvirt.virt_type in
-                ('kvm', 'qemu', 'parallels') and
+                ('kvm', 'qemu', 'parallels', 'ch') and
                 CONF.libvirt.use_virtio_for_bridges):
             model = network_model.VIF_MODEL_VIRTIO
 

